### ISSUE TITLE
adapter-node: add option to specify path to package.json

### ISSUE THREAD

**Issue by Julien-Hery** [sveltejs/kit#11093]
### Describe the problem

I'm using sveltekit in an NX monorepo with a folder structure like this :
```
.
├── ...
├── packages                  
│   ├── some-directory   
│   │   ├── ...
│   │   └── svelte.config.js
└── package.json
```

When I use nx to serve my app I start the command `vite dev` in `packages/some-directory` and the adapter will look for the `package.json` in this directory. The problem is that my package.json is at another level in the tree and I can't run `vite dev` outside my app directory because sveltekit doesn't allow it.

### Describe the proposed solution

I would like to have an option to specify an absolute path to the package.json so that it can be loaded dynamically no matter the directory structure 

```js
export default function (opts = {}) {
  const { out = 'build', precompress, envPrefix = '', polyfill = true, packageJsonPath = 'package.json' } = opts;
  
  return {
    name: '@sveltejs/adapter-node',
    
    async adapt(builder) {
      [...]
    
      const pkg = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
      
      [...]
    }
  }
}
```

And it could be used like that : 

```js
import adapter from '@sveltejs/adapter-node';
import path from 'path';

export default {
  kit: {
    adapter: adapter({
      packageJsonPath: path.resolve('../../package.json')
    })
  }
};
```

### Importance

i cannot use SvelteKit without it

### Additional Information

I have a PR ready that I will link to this issue in case it's accepted 🙂

**Comment by benmccann** [comment]
How do NX and other tools deal with this? Do they specify paths to `package.json` or do they look up the file tree until they find a `package.json`?

How would it ever be useful for multiple SvelteKit projects to share a `package.json`? In almost any real application I'd expect the different SvelteKit projects to have different dependencies making it so that you can't share a `package.json`

**Comment by Julien-Hery** [comment]
@benmccann, I'm not entirely sure about other projects or whether they even check for the package.json.

It's not necessarily about multiple SvelteKit projects; it could be any projects within the monorepo (whether they use Node.js or any other frontend framework). The project I'm currently working on uses microfrontends (with a custom tool), and the repository may contain various elements not directly related to SvelteKit. 
We also use tools like Checkmarx to identify security issues within dependencies. Managing these issues becomes much simpler with shared dependencies in a monorepo (especially with common elements like axios, lodash, ...).
Most, if not all, of our repositories use nx for structuring and maintaining a consistent approach. While we already incorporate Svelte (SPA) in NX, we now require SSR, and I'm considering using SvelteKit for this purpose.

Setting aside the monorepo aspect, based on my experience, it's generally not advisable to enforce a specific directory for a file; it's too restrictive. Allowing consumers to specify a custom path, if necessary, makes things easier for everyone.

**Comment by benmccann** [comment]
> We also use tools like Checkmarx to identify security issues within dependencies. Managing these issues becomes much simpler with shared dependencies in a monorepo (especially with common elements like axios, lodash, ...).

What package manager do you use? If you use `pnpm` you can do `pnpm update -r` to update a dependency in all sub-projects. If you don't declare the dependency in the sub-projects you don't know which ones were actually vulnerable because you don't know which ones were actually using it. This can make remediating this issues far more challenging.

> Setting aside the monorepo aspect, based on my experience, it's generally not advisable to enforce a specific directory for a file; it's too restrictive. Allowing consumers to specify a custom path, if necessary, makes things easier for everyone.

I disagree for a few reasons. One is that adding options makes it harder to understand how to configure SvelteKit. A second is that there are a lot of tools and libraries that deal with `package.json` and if you start putting it in unexpected places then you increase the chances that some of them start breaking or behaving in unexpected ways. Thirdly, it makes it harder for new comers to a codebase to understand that codebase when files can live in multiple different places.

**Comment by Julien-Hery** [comment]
> What package manager do you use?

I use npm; I've heard about pnpm but never had time to check its features (I only used it to follow the contribution steps in my PR). I'll take a look at it 🙂 

> If you don't declare the dependency in the sub-projects, you won't know which ones are actually vulnerable.

In my context, I guess it doesn't really matter; I just want them to not be vulnerable. It's quite easy to find which path is importing a package.

> Adding options makes it harder to understand how to configure SvelteKit.

My suggestion doesn't bring any complexity as it maintains the default behavior. It adds flexibility for people using tools with different structures.

> A lot of tools and libraries deal with package.json.

Honestly, this is the first time I've encountered this issue. In every tool there is usually an option to specify the path to a config file or any other file the project needs to read. Working on such large projects like SvelteKit probably introduces new challenges that I'll never face 😅 

> It's harder for newcomers to a codebase to understand the code when files can live in multiple different places.

I agree with this one, but again, I don't think it's relevant in this situation.

Anyway, I think I will close this issue because I don't believe we will come to an agreement. My issue seems a bit specific to my use case, and as you mentioned on the PR, there are 60+ other PRs waiting that are probably more important 😄 

Thank you for taking a look at this; have a nice day 🙂
